docs/src/pages/ComponentArchitecture.mdx

```diff
@@ -0,0 +1,97 @@
+export const meta = {
+  title: 'Component Architecture',
+}
+
+The design and architecture of components should follow from a set of first principles.
+
+- There is 1 right way to do things, and it's easy to do the right thing.
+- Consumers should not be penalized for code they don't use (through size, perf, etc).
+- Don't overgeneralize until we have a clear need (abstracting should solve a bundle size goal, or DRY goal, or partner use case)
+
+# Base components
+
+Each component should be built upon a "base component".
+
+In order to keep sorted lists formatted neatly, each base component should be named with the form: `ComponentNameBase.tsx`
+
+#### Applications which load their own CSS in other means should not pay the penalty of loading a runtime CSS-in-JS package, especially if the class names are known at time of authoring.
+
+Base components should expect a prop called `classes`, which is either a
+dictionary of keys to classnames (where a key represents a portion of the DOM), or
+a callback which will return the same dictionary when called with the component's
+state.
```

Comment:
#### Do not force to use any CSS or CSS-in-JS implementation

Applications which load their own CSS in other means should not pay the penalty of loading a runtime CSS-in-JS package, especially if the class names are known at time of authoring.

#### A component should be able to render any theme (including tokens, styles, and icons) or any CSS

???

#### Theming is first-class citizen

A proposed CSS-in-JS solution should allow themes to cascade and merge allowing subtree themes to be updated at runtime and have the ability to adjust the incoming theme.

#### RTL mode should be configurable for the entire application or a subtree of the application

```tsx
// 👍, `Provider` is an abstract name
<Provider direction='rtl' />
// 👎 
<>
  <Button direction='rtl' />
  <Status direction='rtl' />
</>
```

Reply:
Added the last one, though I would like to discuss the first 3 in another PR.

Reply:
The first point on CSS-in-JS solutions probably falls under our more broad requirement/philosophy:

> ### The consumer should be able to override, extend, or replace any opinions of the library.
> State ...
> styling ...
> A11y ...
> Slots ..
> Styling approach (css, css-in-js, etc) ...
> css-in-js renderer ...

Something like this.  It may be done through composition or flexible components but the need to change the essence of the component is what we're targeting.